File: src/utils/graph_generator.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <random>
#include <limits>

using namespace std;

const int INF = 1e9;

vector< vector<int> > generateGraphFromText(const string &filename) {
    ifstream infile(filename);
    if (!infile) {
        cerr << "Error opening file" << endl;
        return vector< vector<int> > ();
    }

    vector< vector<int> > graph;
    string line;
    
    while (getline(infile, line)) {
        vector<int> row;
        string weight;
        istringstream iss(line);
        
        while (iss >> weight) {
            if (weight == "INF") {
                row.push_back(INF);
            } else {
                row.push_back(stoi(weight));
            }
        }
        
        graph.push_back(row);
    }

    infile.close();
    return graph;
}

vector< vector<int> > generateRandomGraph(int V, double density){   // 0 <= density < 1
    srand(time(NULL));
    vector< vector<int> > graph(V, vector<int> (V));
    for(int i=0;i<V;i++){
        for(int j=0;j<V;j++){
            int weight = rand()%INF;
            if(weight > INF*density) graph[i][j] = INF;
            else graph[i][j] = weight;
        }
    }
    return graph;
}

void printMatrix(const vector< vector<int> >& dist) {
    int V = dist.size();
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (dist[i][j] == INF)
                cout << "INF ";
            else
                cout << dist[i][j] << " ";
        }
        cout << endl;
    }
}

int main(){
    cout << "Generating Random Graph:\n";
    vector< vector<int> > randGraph = generateRandomGraph(4, 0.75);
    printMatrix(randGraph);

    cout << "\n";

    cout << "Generating Graph from Text File:\n";
    string filename = "data/input/4x4.txt";
    vector< vector<int> > textGraph = generateGraphFromText(filename);
    
    if (textGraph.empty()) {
        cerr << "Failed to load graph from file: " << filename << endl;
        return 1;
    }

    printMatrix(textGraph);
    return 0;
}

File: src/utils/graph_generator.js
const fs = require("fs");
const readline = require("readline");

const INF = 1e9;

function generateGraphFromText(filename) {
  return new Promise((resolve, reject) => {
    const graph = [];
    const rl = readline.createInterface({
      input: fs.createReadStream(filename),
      crlfDelay: Infinity,
    });

    rl.on("line", (line) => {
      const row = line
        .split(" ")
        .map((weight) => (weight === "INF" ? INF : parseInt(weight, 10)));
      graph.push(row);
    });

    rl.on("close", () => resolve(graph));
    rl.on("error", (err) => reject(err));
  });
}

function generateRandomGraph(V, density) {
  const graph = Array.from({ length: V }, () => Array(V).fill(0));

  for (let i = 0; i < V; i++) {
    for (let j = 0; j < V; j++) {
      if (i == j) {
        graph[i][j] = 0;
        continue;
      }
      const weight = Math.floor(Math.random() * INF);
      graph[i][j] = weight > INF * density ? INF : weight;
    }
  }

  return graph;
}

function printGraph(graph) {
  graph.forEach((row) => {
    console.log(
      row.map((weight) => (weight === INF ? "INF" : weight)).join(" ")
    );
  });
}

async function main() {
  console.log("Generating Random Graph:");
  const randGraph = generateRandomGraph(4, 0.75);
  printGraph(randGraph);
  console.log("--------");

  console.log("Generating Graph from Text File:");
  const filename = "data/input/4x4.txt";

  try {
    const textGraph = await generateGraphFromText(filename);
    printGraph(textGraph);
  } catch (err) {
    console.error(`Failed to load graph from file: ${filename}`);
    console.error(err);
  }
  console.log("--------");
}

main();
